### DISCUSSION TITLE
SVG as vector graphic in PDF?

### DISCUSSION THREAD

**Discussion by hvbtup** [eclipse-birt/birt#1192] [General]
IIRC BIRT 4.3.0 supported rendering SVG images (Image items, not background-image) as vector graphic with the PDF emitter.

Yesterday I tested images with the master branch and now it seems that SVGs are always converted to bitmaps in the PDF emitter.

Apart from sizing issues (which I'll describe in another discussion), this has a disadvantage:

With a DPI resolution good for printing (e.g. 300 or 600), this results in bigger PDF files.
With a lower DPI resolution, this results in visible pixels in printed output.

**Comment by speckyspooky** [comment]
I can confirm that there are any topics of SVG usage (I work with BIRT 4.6 & 4.8).
The could shows accordinh changes but I haven#T found some hints/documentation why there are such changes of the usage of SVG.

So we are not able to add SVG realy in a simple way to the report but some code snippets show that this was planed.
I also tried to you to use SVG on PDF because we get so much better document quality with the SVG-images.

Nevertheless I think we need time to investigate on it and had to change the image handling parts.

**Comment by hvbtup** [comment]
I just tested to create a PDF with an SVG image in BIRT 4.2.1.
I can now confirm that this used to work with BIRT 4.2.1 and BIRT 4.3.0.
The resulting PDF is 7.7 KB.

The result is a PDF without an image - the SVG drawing instructions result in PDF vector drawing instructions (not a bitmap).
To verify this, it is best to use Adobe Reader and zoom to 1000%.

So this very useful functionality somehow got lost between 4.3.0 and 4.10.

Can you test my report with BIRT 4.6 and 4.8?
[svg_image.zip](https://github.com/eclipse/birt/files/10533649/svg_image.zip)

**Comment by hvbtup** [comment]
[svg-test.zip](https://github.com/eclipse/birt/files/10534241/svg-test.zip)

I created and ran another test (see attachment svg-test.zip) for SVG images with the PDF emitter.

SVG images as vector graphics actually *do* work in many cases.

To see if an image is a bitmap or vector graphics, it is best to open the PDF in Adobe Reader instead of the browser.
 - Bitmap images can be selected (as a rectangle) and will show pixels in high zoom levels.
 - Vector graphics cannot be selected and look sharp at every zoom level.

SVG image item, source: URI: OK
SVG image item, source: shared resource, only filename specified: bug 1
SVG image item, source: shared resource, using file:/... URI syntax: OK
SVG background item: bug 2, remark 3

Bug 1:
When the same image was used as a background image before, the image is rendered as a bitmap and the proportions are wrong.
I don't know if exactly this fact triggers the error, but that's what I see in the test report.

Bug 2:
The image is rendered as a bitmap instead of vector graphics.

Remark 3:
The bitmap seems to use 96 DPI (so the tiger head bitmap is ~ 20 cm wide).
This is not wrong, because e.g. Inkscape and browsers assume 96 dpi (other software uses different values).
But an SVG file doesn't really have a physical size.
My upcoming draft PR will change this behavior. I uses the reports' imageDpi property (300 in the example report), so the SVG is still rendered as a bitmap, but much smaller and finer. The PR also allows to override this setting.
However, I am not sure if this is the right way to go: We have the background-image-height and background-image-width properties, which seem to be ignored completely. It would probably be better to use these values to set an absolute size or eg. 100% (of the containing item's width and height).

**Comment by hvbtup** [comment]
See draft PR #1195, which doesn't fix the "PDF is rendered as bitmap" issue, but at least allows to set a different DPI value for print-quaility bitmaps from SVG backround images.

**Comment by hvbtup** [comment]
@speckyspooky You recently looked into the details of image handling, so you probably know this stuff better than I do: Are you able to fix (some of) the issues and remarks I mentioned?

**Reply by hvbtup** [reply]
Hmm... I searched for references of `getBackgroundWidth` and the only call I found (ignoring other emitters) was in `....nLayout.area.impl.PageArea`. So it seems that these properties are only used for the *page* background image, but not for other background images.

**Reply by speckyspooky** [reply]
@hvbtup I tested your test example currently with file-URI and thsi case work with BIRT 4.6, 4.10, 4.12
BIRT 4.8 gives an internal and doesn't open the report file.
The created PDF processed correctly with embedded SVG-images.

Yes, the SVG-images are noted on my list. Currently I work on another topic and afterwards I will switch to it and will check some details.
Thanks for your verification this will help to figure out the details.
